### PR TITLE
DPL-990 - Modify pacbio libraries to support aliquots

### LINKED ISSUE #1154: DPL-990 - As developers we need to modify libraries to use aliquots to support revio multiplexing
**User story**
As developers we need to modify libraries to use aliquots to support revio multiplexing.

This endpoint should be feature flagged until we are ready to migrate single-plexed pool to libraries.

**Who are the primary contacts for this story**
James / Steve

**Who is the nominated tester for UAT**
PSD

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] A library can have many aliquots
- [ ] New endpoint to support library creation
- [ ] Support library creation without pools.
- [ ] Migration to convert single-plexed pools to libraries.

**Dependencies**
This story is blocked by the following dependencies:
- #<issue_no.>
- sanger/<repo>#<issue_no.>

References
This story has a non-blocking relationship with:
- #1137
- sanger/<repo>#<issue_no.>

**Additional context**

### PR DESCRIPTION
**Pull request by BenTopping**
Closes #1154

#### Changes proposed in this pull request

- Pacbio Libraries can have aliquots
- Rake task to migrate libraries to aliquots

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

**Review by stevieing — CHANGES_REQUESTED** [review]
I have only set it to request changes to open a discussion rather than there being changes that need to be done. I think it would be good to try and centralise any behaviour related to aliquots wherever possible.

File: app/models/pacbio/library.rb [hunk 1]
```diff
@@ -35,6 +36,24 @@ class Library < ApplicationRecord
 
     has_one :source_plate, through: :source_well, source: :plate, class_name: '::Plate'
 
+    # TODO: remove pool constraint this when pools are updated for aliquots
+    validates :primary_aliquot, presence: true, if: -> { pool.blank? }
+    accepts_nested_attributes_for :primary_aliquot, allow_destroy: true
+
+    after_create :create_used_aliquot, if: -> { pool.blank? }
+
+    def create_used_aliquot
```
[the remark this answers is not in the document]
Comment (BenTopping) [reply app/models/pacbio/library.rb:45]: Yeah I think its possible after we implement aliquots for pools and wells. I dont want to create any premature aliquot data for them so I think it is safer to leave that in here for the time being. We also have libraries from pools to deal with for the time being so we need this 'if no pool' condition.

File: lib/tasks/migrate_pacbio_aliquot_data.rake [hunk 2]
```diff
@@ -30,8 +30,39 @@ namespace :pacbio_aliquot_data do
...
   end
 
+  task migrate_library_data: :environment do
+    puts '-> Creating primary aliquots for all libraries and derived aliquots for all requests used in libraries'
+
+    Pacbio::Pool.find_each do |pool|
+      next unless pool.libraries.count == 1
+
+      library = pool.libraries.first
+      new_lib = library.dup
```
[the remark this answers is not in the document]
Comment (BenTopping) [reply lib/tasks/migrate_pacbio_aliquot_data.rake:40]: I suspect I was trying to test something and just abbreviated it quickly and left it. Will update.

**Review by stevieing — APPROVED** [review]

**Comment by BenTopping** [comment]
Close in favour of https://github.com/sanger/traction-service/pull/1219

**Pull request closed without merging**